crates/pera-core/src/signature_mpc/mpc_manager.rs

```diff
@@ -0,0 +1,450 @@
+                debug!("event: Finalize MPC Session {:?}", event);
+            };
+        }
+        Ok(())
+    }
+
+    /// Advance all the MPC instances that either received enough messages to, or perform the first step of the flow.
+    /// We parallelize the advances with Rayon to speed up the process.
+    pub async fn handle_end_of_delivery(&mut self) -> PeraResult {
+        let mut ready_to_advance = self
+            .mpc_instances
+            .iter_mut()
+            .filter_map(|(_, instance)| {
+                let threshold_number_of_parties = ((self.number_of_parties * 2) + 2) / 3;
+                if (instance.status == MPCSessionStatus::Active
+                    && instance.pending_messages.len() >= threshold_number_of_parties)
+                    || (instance.party.is_none() && instance.status == MPCSessionStatus::Active)
+                {
+                    Some(instance)
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<&mut SignatureMPCInstance<P>>>();
+        ready_to_advance
+            .par_iter_mut()
+            // TODO (#263): Mark and punish the malicious validators that caused some advances to return None, a.k.a to fail
+            .map(|ref mut instance| instance.advance(&self.auxiliary_input))
+            .collect::<PeraResult<_>>()?;
+        Ok(())
+    }
+
+    /// Handles a message by forwarding it to the relevant MPC instance
+    /// If the instance does not exist, punish the sender
+    pub fn handle_message(
+        &mut self,
+        message: &[u8],
+        authority_name: AuthorityName,
+        session_id: ObjectID,
+    ) -> PeraResult<()> {
+        let Some(mut instance) = self.mpc_instances.get_mut(&session_id) else {
+            error!(
+                "received a message for instance {:?} which does not exist",
+                session_id
+            );
+            // TODO (#261): Punish a validator that sends a message related to a non-existing mpc instance
+            return Ok(());
+        };
+        instance.handle_message(SignatureMPCMessage {
+            message: message.to_vec(),
+            authority: authority_name,
+        })
+    }
+
+    /// Spawns a new MPC instance if the number of active instances is below the limit
+    /// Otherwise, adds the instance to the pending queue
+    fn push_new_mpc_instance(&mut self, event: P::InitEvent) {
+        if self.mpc_instances.contains_key(&event.session_id().bytes) {
+            // This should never happen, as the session ID is a move UniqueID
+            error!(
+                "Received start flow event for session ID {:?} that already exists",
+                event.session_id()
+            );
+            return;
+        }
+
+        info!(
+            "Received start flow event for session ID {:?}",
+            event.session_id()
+        );
+
+        if self.active_instances_counter > self.max_active_mpc_instances {
```

Comment:
@ycscaly when a Validator synchronises after being down, they should listen for Init and Finished events from the Move
VM that happened because of transactions while while they were away and increase and decrease the counter accordingly.
This should make sure that they are back in sync.

@dolevmu if the assumption that at least two thirds of the stake is honest and alive, then there cannot be old instances
that "never advance for some reason". To be safe, we could implement a timeout (15 minutes or something), but it should
never be reached if the assumption holds.

Reply:
@ItayLevyOfficial

Reply:
I agree with @cdesaintguilhem on this one. There's an [existing issue](https://github.com/dwallet-labs/dwallet-network/pull/280/files#r1820448440) on implementing new validators sync

Reply:
`when a Validator synchronises after being down, they should listen for Init and Finished events f` - is this tracked?

Reply:
This will be implemented as part of the [validator synchronization](https://github.com/dwallet-labs/dwallet-network/issues/311) issue